### PR TITLE
Replaced some unwrap_or and map_or with lazy variants

### PR DESCRIPTION
**Pull request by klensy**
Replaced some `unwrap_or` and `map_or` with `unwrap_or_else` and `map_or_else`.

### CODE REVIEW

**Comment by klensy** [comment]
There's a bug in rust-log-analyzer, it shows:
```
-            .self_desc
-            .clone()
-            .clone()
-            .map_or_else(|| String::new(), |ty| format!(" for type `{}`", ty))
+        overlap.self_desc.clone().map_or_else(|| String::new(), |ty| format!(" for type `{}`", ty))
```

but here actually:
```
-        overlap
-            .self_desc
-            .clone()
-            .map_or_else(|| String::new(), |ty| format!(" for type `{}`", ty))
+        overlap.self_desc.clone().map_or_else(|| String::new(), |ty| format!(" for type `{}`", ty))
```

File: compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs [hunk 1]
```diff
@@ -2372,7 +2372,7 @@ fn compute_type_parameters(cx: &CodegenCx<'ll, 'tcx>, ty: Ty<'tcx>) -> &'ll DIAr
     fn get_parameter_names(cx: &CodegenCx<'_, '_>, generics: &ty::Generics) -> Vec<Symbol> {
         let mut names = generics
             .parent
-            .map_or(vec![], |def_id| get_parameter_names(cx, cx.tcx.generics_of(def_id)));
+            .map_or_else(|| vec![], |def_id| get_parameter_names(cx, cx.tcx.generics_of(def_id)));
```
Comment (estebank) [inline compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:2375]:
```suggestion
            .map_or_else(Vec::new, |def_id| get_parameter_names(cx, cx.tcx.generics_of(def_id)));
```

File: compiler/rustc_codegen_llvm/src/debuginfo/mod.rs [hunk 2]
```diff
@@ -481,9 +481,10 @@ impl DebugInfoMethods<'tcx> for CodegenCx<'ll, 'tcx> {
         }
 
         fn get_parameter_names(cx: &CodegenCx<'_, '_>, generics: &ty::Generics) -> Vec<Symbol> {
-            let mut names = generics
-                .parent
-                .map_or(vec![], |def_id| get_parameter_names(cx, cx.tcx.generics_of(def_id)));
+            let mut names = generics.parent.map_or_else(
+                || vec![],
```
Comment (estebank) [inline compiler/rustc_codegen_llvm/src/debuginfo/mod.rs:485]:
```suggestion
                Vec::new,
```

File: compiler/rustc_codegen_llvm/src/metadata.rs [hunk 3]
```diff
@@ -65,8 +65,8 @@ fn search_meta_section<'a>(
         while llvm::LLVMIsSectionIteratorAtEnd(of.llof, si.llsi) == False {
             let mut name_buf = None;
             let name_len = llvm::LLVMRustGetSectionName(si.llsi, &mut name_buf);
-            let name = name_buf.map_or(
-                String::new(), // We got a NULL ptr, ignore `name_len`.
+            let name = name_buf.map_or_else(
+                || String::new(), // We got a NULL ptr, ignore `name_len`.
```
Comment (estebank) [inline compiler/rustc_codegen_llvm/src/metadata.rs:69]:
```suggestion
                String::new, // We got a NULL ptr, ignore `name_len`.
```

File: compiler/rustc_mir/src/const_eval/eval_queries.rs [hunk 4]
```diff
@@ -50,7 +50,7 @@ fn eval_body_using_ecx<'mir, 'tcx>(
 
     let name =
         with_no_trimmed_paths(|| ty::tls::with(|tcx| tcx.def_path_str(cid.instance.def_id())));
-    let prom = cid.promoted.map_or(String::new(), |p| format!("::promoted[{:?}]", p));
+    let prom = cid.promoted.map_or_else(|| String::new(), |p| format!("::promoted[{:?}]", p));
```
Comment (estebank) [inline compiler/rustc_mir/src/const_eval/eval_queries.rs:53]:
```suggestion
    let prom = cid.promoted.map_or_else(String::new, |p| format!("::promoted[{:?}]", p));
```

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 5]
```diff
@@ -223,7 +223,7 @@ impl<'a> Parser<'a> {
         fn tokens_to_string(tokens: &[TokenType]) -> String {
             let mut i = tokens.iter();
             // This might be a sign we need a connect method on `Iterator`.
-            let b = i.next().map_or(String::new(), |t| t.to_string());
+            let b = i.next().map_or_else(|| String::new(), |t| t.to_string());
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/diagnostics.rs:226]:
```suggestion
            let b = i.next().map_or_else(String::new, |t| t.to_string());
```

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 6]
```diff
@@ -1971,65 +1971,68 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
-                        Scope::Root => break false,
-
-                        Scope::Body { .. } => break true,
-
-                        Scope::Binder { s, .. }
-                        | Scope::Elision { s, .. }
-                        | Scope::ObjectLifetimeDefault { s, .. } => {
-                            scope = s;
+        let object_lifetime_defaults = type_def_id.map_or_else(
+            || vec![],
```
Comment (estebank) [inline compiler/rustc_resolve/src/late/lifetimes.rs:1975]:
```suggestion
            Vec::new,
```

File: compiler/rustc_session/src/filesearch.rs [hunk 7]
```diff
@@ -169,7 +169,7 @@ pub fn get_or_default_sysroot() -> PathBuf {
 
     // Check if sysroot is found using env::args().next(), and if is not found,
     // use env::current_exe() to imply sysroot.
-    from_env_args_next().unwrap_or(from_current_exe())
+    from_env_args_next().unwrap_or_else(|| from_current_exe())
```
Comment (estebank) [inline compiler/rustc_session/src/filesearch.rs:172]:
```suggestion
    from_env_args_next().unwrap_or_else(from_current_exe)
```

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 8]
```diff
@@ -349,7 +349,7 @@ fn report_negative_positive_conflict(
         E0751,
         "found both positive and negative implementation of trait `{}`{}:",
         overlap.trait_desc,
-        overlap.self_desc.clone().map_or(String::new(), |ty| format!(" for type `{}`", ty))
+        overlap.self_desc.clone().map_or_else(|| String::new(), |ty| format!(" for type `{}`", ty))
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:352]:
```suggestion
        overlap.self_desc.clone().map_or_else(String::new, |ty| format!(" for type `{}`", ty))
```

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 9]
```diff
@@ -397,7 +397,10 @@ fn report_conflicting_impls(
         let msg = format!(
             "conflicting implementations of trait `{}`{}:{}",
             overlap.trait_desc,
-            overlap.self_desc.clone().map_or(String::new(), |ty| { format!(" for type `{}`", ty) }),
+            overlap
+                .self_desc
+                .clone()
+                .map_or_else(|| String::new(), |ty| { format!(" for type `{}`", ty) }),
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:403]:
```suggestion
                .map_or_else(String::new, |ty| { format!(" for type `{}`", ty) }),
```

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 10]
```diff
@@ -415,7 +418,9 @@ fn report_conflicting_impls(
                     impl_span,
                     format!(
                         "conflicting implementation{}",
-                        overlap.self_desc.map_or(String::new(), |ty| format!(" for `{}`", ty))
+                        overlap
+                            .self_desc
+                            .map_or_else(|| String::new(), |ty| format!(" for `{}`", ty))
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:423]:
```suggestion
                            .map_or_else(String::new, |ty| format!(" for `{}`", ty))
```

File: compiler/rustc_typeck/src/check/method/probe.rs [hunk 11]
```diff
@@ -1695,7 +1695,7 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
             } else {
                 self.fcx
                     .associated_item(def_id, name, Namespace::ValueNS)
-                    .map_or(Vec::new(), |x| vec![x])
+                    .map_or_else(|| Vec::new(), |x| vec![x])
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/method/probe.rs:1698]:
```suggestion
                    .map_or_else(Vec::new, |x| vec![x])
```

File: compiler/rustc_typeck/src/check/mod.rs [hunk 12]
```diff
@@ -1062,7 +1062,10 @@ fn report_unexpected_variant_res(tcx: TyCtxt<'_>, res: Res, span: Span) {
         E0533,
         "expected unit struct, unit variant or constant, found {}{}",
         res.descr(),
-        tcx.sess.source_map().span_to_snippet(span).map_or(String::new(), |s| format!(" `{}`", s)),
+        tcx.sess
+            .source_map()
+            .span_to_snippet(span)
+            .map_or_else(|_| String::new(), |s| format!(" `{}`", s)),
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/mod.rs:1068]:
```suggestion
            .map_or_else(String::new, |s| format!(" `{}`", s)),
```

File: compiler/rustc_typeck/src/check/pat.rs [hunk 13]
```diff
@@ -879,7 +879,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let sm = tcx.sess.source_map();
             let path_str = sm
                 .span_to_snippet(sm.span_until_char(pat.span, '('))
-                .map_or(String::new(), |s| format!(" `{}`", s.trim_end()));
+                .map_or_else(|_| String::new(), |s| format!(" `{}`", s.trim_end()));
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/pat.rs:882]:
```suggestion
                .map_or_else(String::new, |s| format!(" `{}`", s.trim_end()));
```

File: compiler/rustc_typeck/src/collect.rs [hunk 14]
```diff
@@ -2387,7 +2387,7 @@ fn compute_sig_of_foreign_fn_decl<'tcx>(
                     .sess
                     .source_map()
                     .span_to_snippet(ast_ty.span)
-                    .map_or(String::new(), |s| format!(" `{}`", s));
+                    .map_or_else(|_| String::new(), |s| format!(" `{}`", s));
```
Comment (estebank) [inline compiler/rustc_typeck/src/collect.rs:2390]:
```suggestion
                    .map_or_else(String::new, |s| format!(" `{}`", s));
```

**Review by estebank — COMMENTED** [review]
You can always pass any callable to a place that expects a closure, as long as the number of arguments matches. I'm surprised we don't lint about this.

File: compiler/rustc_typeck/src/check/mod.rs (hunk 12 above)
Comment (klensy) [reply compiler/rustc_typeck/src/check/mod.rs:1068]: >as long as the number of arguments matches

You catched me here and with next 2 suggestions )

File: compiler/rustc_typeck/src/collect.rs (hunk 14 above)
Comment (osa1) [reply compiler/rustc_typeck/src/collect.rs:2390]: This was marked as resolved but it's not resolved.

Comment (osa1) [reply compiler/rustc_typeck/src/collect.rs:2390]: Ah, this is `Result`, not `Option`, nvm.

**Pull request merged**